components/OlasTokenPage/SupplyPieChart.jsx

```diff
@@ -0,0 +1,200 @@
+
+      const olasContract = new web3.eth.Contract(olasAbi, olasAddress);
+
+      const promises = [
+        olasContract.methods.totalSupply().call(),
+        ...ADDRESSES.map((address) => olasContract.methods.balanceOf(address).call()),
+      ];
+
+      const result = await Promise.allSettled(promises);
+
+      const totalSupplyResult = result[0].value;
+      const distributions = result
+        .slice(1, result.length)
+        .map((item) => item.value || 0);
+
+      const circulatingSupply = totalSupplyResult > 0
+        ? result[0].value
+            - distributions.reduce((sum, item) => sum + item, 0n)
+        : 0;
+
+      setTotalSupply(formatEthers(totalSupplyResult));
+      setData([
+        ...distributions.map((item) => formatEthers(item)),
+        formatEthers(circulatingSupply),
+      ]);
+
+      setLoading(false);
+    };
+
+    try {
+      fetchData();
+    } catch (error) {
+      console.error('Error fetching data:', error);
+    }
+  }, []);
+
+  return (
+    <>
+      <div className="flex flex-wrap gap-8 p-4 border-b">
+        <div>
+          <h2 className="text-sm text-slate-500 font-bold tracking-widest uppercase">
+            Total Supply
+          </h2>
+          <div className="text-gradient text-4xl font-extrabold">
+            {loading ? '--' : formatNumber(totalSupply)}
+          </div>
+          <div className="mb-4">
+            <Verify url="https://etherscan.io/address/0x0001A500A6B18995B03f44bb040A5fFc28E45CB0#readContract#F16" />
+          </div>
+        </div>
+        <div>
+          <h2 className="text-sm text-slate-500 font-bold tracking-widest uppercase">
+            Circulating Supply
+          </h2>
+          <div className="text-gradient text-4xl font-extrabold">
+            {loading ? '--' : formatNumber(data[data.length - 1])}
+          </div>
+          <div className="mb-4">
+            <Verify url="https://www.coingecko.com/en/coins/autonolas" />
+          </div>
+        </div>
+      </div>
+      <div className="flex flex-col p-4">
+        <h2 className="text-sm text-slate-500 font-bold tracking-widest uppercase mb-4">
+          Total Supply Distribution
+        </h2>
+        <div className="flex flex-col items-center gap-8">
+          <div>
+            {loading ? (
+              <div className="text-center">Loading...</div>
+            ) : (
+              <Pie
+                data={{
+                  labels: LABELS,
+                  datasets: [
+                    {
+                      data,
+                      backgroundColor: RGB_COLOR,
+                      hoverBackgroundColor: RGB_COLOR,
+                    },
+                  ],
+                }}
+              />
+            )}
+          </div>
+
+          <div className="flex flex-col gap-2 self-center w-full">
+            {data.map((item, index) => (
+              <LegendItem
+                key={index}
```

Comment:
should avoid `key={index}` and use unique ids specific to the data.
may need to change the data once it's fetched, and for example, make it an object with identifier

shouldn't cause any issues though as it's not being filtered or sorted

Reply:
updated